### PR TITLE
refactor(crypto): split Ecdh out of EcdhMask

### LINKED ISSUE #879: dev: split the Jubjub key agreement Ecdh out of EcdhMask
`crypto/EcdhMask` carries the Jubjub key agreement inline, so a consumer that needs several masks from one shared secret (the `Field` pad in #735) cannot reach the shared point without going through `encrypt`.

Move the key agreement into its own `crypto/Ecdh` module: `deriveShared` (sender side, owns the identity-key and zero-ephemeral guards), `recoverShared` (recipient side), and a `SharedSecret` struct. `EcdhMask.encrypt` / `decrypt` delegate to it and keep their ciphertexts unchanged.

Prerequisite for #735.

PR: #866

### PR DESCRIPTION
**Pull request by 0xisk**
## Types of changes

What types of changes does your code introduce to OpenZeppelin Midnight Contracts?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update (if none of the other choices apply)

Closes #879
Refs #735 (prerequisite). The `Field` pad is stacked on this branch in #913.

Splits the Jubjub key agreement out of `crypto/EcdhMask` into `crypto/Ecdh` so a consumer can derive several masks from one shared secret without going through `encrypt`. `encrypt` output is unchanged; the golden-vector test pins it.

Not visible in the diff:

- The weak-input guards move with the key agreement, so the assert text is now `Ecdh: identity pk` / `Ecdh: zero ephemeral`. Nothing in the repo matched on the old `EcdhMask:` prefix.
- `EcdhMask` does not re-export the `Ecdh` names. A re-exported import keeps the source module's prefix (`EcdhMask_Ecdh_deriveShared`), so consumers import both modules instead.
- Circuits are ordered caller then callee, which is why `kdf` moved below `encrypt`.

## PR Checklist

- [x] I have read the [Contributing Guide](../CONTRIBUTING.md#your-first-code-contribution)
- [x] I have added tests that prove my fix is effective or that my feature works. See [GUIDELINES.md#testing](../GUIDELINES.md#testing) for more information.
- [x] I have added documentation for new methods or changes to existing method behavior. See [GUIDELINES.md#documentation](../GUIDELINES.md#documentation) for more information.
- [x] CI Workflows Are Passing

## Summary by CodeRabbit

* **New Features**
  * Added stateless Jubjub Diffie-Hellman key agreement APIs for deriving and recovering shared secrets.
  * Added support for ephemeral public keys and shared-point outputs.
  * Preserved ciphertext compatibility for existing ECDH-based masking and encryption.

* **Bug Fixes**
  * Improved validation and error messages for invalid identity keys and zero ephemeral values.

* **Tests**
  * Added coverage for key agreement, recovery, edge cases, invalid inputs, and fixed encryption output vectors.

### CODE REVIEW

File: contracts/src/crypto/test/mocks/MockEcdh.compact
Comment (andrew-fleming) [inline contracts/src/crypto/test/mocks/MockEcdh.compact:1]: Should we not make these impure for testing? Or do you think it's better to wait for #850?

File: contracts/src/crypto/test/Ecdh.test.ts
Comment (andrew-fleming) [inline contracts/src/crypto/test/Ecdh.test.ts:1]: Some additional tests worth considering:

- The idea of this split if I understand correctly is that a consumer can derive several masks from one shared secret without going through encrypt. We should check that the decomposed path agrees with the encrypt path
- Test the module's boundary with mixed-order and off-curve through deriveShared (recipientPk) and recoverShared (ephemeralPk)
- DH symmetry -> `deriveShared(g^a, b).sShared === deriveShared(g^b, a).sShared`

File: contracts/src/crypto/Ecdh.compact [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+
+  /**
+   * @description Runs the sender's ECDH step: the ephemeral public key `g^e` to
+   * publish and the shared secret `pk^e` to derive from. The single site of the
+   * weak-input guards.
+   *
+   * @notice `e` MUST be fresh per call and secret (see the module freshness
+   * precondition).
+   *
+   * @circuitInfo k=11, rows=1385
```
Comment (andrew-fleming) [inline contracts/src/crypto/Ecdh.compact:95]: Just curious, are these numbers for v2 or v3?

**Review by andrew-fleming — COMMENTED** [review]
The split looks clean, @0xisk! I left a couple of questions and a suggestion for a few other tests

File: contracts/src/crypto/test/Ecdh.test.ts
Comment (0xisk) [reply contracts/src/crypto/test/Ecdh.test.ts:1]: Agree — fixed in ca10fb21 (symmetry + boundary in Ecdh.test.ts, composition-equals-encrypt in EcdhMask.test.ts).

File: contracts/src/crypto/test/mocks/MockEcdh.compact
Comment (0xisk) [reply contracts/src/crypto/test/mocks/MockEcdh.compact:1]: Agree — waiting for #850. Once it lands I rebase and convert MockEcdh the same way (`_invocations` + simulator).

File: contracts/src/crypto/Ecdh.compact (hunk 1 above)
Comment (0xisk) [reply contracts/src/crypto/Ecdh.compact:96]: v3. `compile:crypto` runs with `--feature-zkir-v3` on this branch, and MockEcdh is not in the v2 exclude list #850 adds, so the numbers stay v3.

File: contracts/src/crypto/Ecdh.compact [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+ *   `secretToScalar`); a shared key module is its natural home.
+ *
+ * @dev Freshness and secrecy precondition (CRITICAL). `e` MUST be both unique per
+ * call to the same recipient AND secret (unpredictable to observers), even across
+ * circuit invocations that reuse a randomness seed.
+ *
+ * Uniqueness: a repeated `e` to one recipient repeats `S`, so every key, mask or
+ * nonce a consumer derives from `S` repeats too. Where that derivation is a
+ * one-time pad (`crypto/EcdhMask`) the reuse is catastrophic: subtracting two
+ * ciphertexts recovers both plaintexts.
```
[the remark this answers is not in the document]
Comment (0xisk) [reply contracts/src/crypto/Ecdh.compact]: Done in f244aee9 — the doc now states the plaintext difference leaks, and either plaintext once the other is known or constrained.
